ledger-core/network/nds/msgdelivery/ctl_retry_msg_worker.go

```diff
@@ -23,26 +23,30 @@ func newRetryMsgWorker(sender *msgSender, parallel, postponed int) *retryMsgWork
 	return &retryMsgWorker{
 		sender:    sender,
 		sema:      synckit.NewSemaphore(parallel),
-		marks:     nodeMarks{marks: make(map[uint32]struct{}, parallel)},
-		postponed: make([]postponedMsg, postponed),
+		postponed: make(chan postponedMsg, postponed),
+		marks:     &sender.marks,
 	}
 }
 
 type retryMsgWorker struct {
 	sender *msgSender
 	sema   synckit.Semaphore
-	marks  nodeMarks
+	marks  *nodeMsgMarks
 
-	// circular buffer
-	postponed   []postponedMsg
-	read, write int32
+	// buffer
+	postponed     chan postponedMsg
+	postponedRead <- chan postponedMsg
+	postponedLimit uint32
 
 	wakeup chan struct{}
 }
 
+const minRetryBulkSize = 32
```

Comment:
What is the heuristic constant? Should it be configurable?

Reply:
to get a noticeable difference between n*repeatFn vs bulkFn
low value will cause mostly empty arrays to stuck in stages, high value will increase CPU usage

doesn't need to be configured as almost any value is tolerable